File: src/lib/paths.ts
import path from "node:path";
import fs from "node:fs";

/**
 * All mutable runtime state lives under the data dir. Relocatable via
 * WB_DATA_DIR so a packaged desktop app can write to the user-data folder
 * (the app bundle itself is read-only). Defaults to ./data for server/dev.
 */
export const DATA_DIR = process.env.WB_DATA_DIR || path.join(process.cwd(), "data");
export const CONFIG_PATH = path.join(DATA_DIR, "config.json");
export const LISTINGS_PATH = path.join(DATA_DIR, "listings.json");
export const QUEUE_PATH = path.join(DATA_DIR, "queue.json");
export const IMAGES_DIR = path.join(DATA_DIR, "images");

export function ensureDataDirs() {
  for (const dir of [DATA_DIR, IMAGES_DIR]) {
    if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  }
}

/** Atomic write: write to a temp file then rename (same-dir rename is atomic),
 * so a crash/power-loss mid-write can't truncate the real file. */
export function atomicWrite(file: string, data: string) {
  ensureDataDirs();
  const tmp = `${file}.tmp.${process.pid}`;
  fs.writeFileSync(tmp, data);
  fs.renameSync(tmp, file);
}

/** Move a corrupt/unparseable file aside instead of silently overwriting it,
 * so existing data (keys, history) is recoverable. */
export function quarantineCorrupt(file: string) {
  try {
    if (fs.existsSync(file)) fs.renameSync(file, `${file}.corrupt-${Date.now()}`);
  } catch {
    /* ignore */
  }
}

File: src-tauri/src/paths.rs
#![allow(dead_code)]
//! Runtime data dir + atomic writes (ported from src/lib/paths.ts).
//! In the desktop app the base dir is Tauri's per-user app_data_dir.

use std::fs;
use std::path::{Path, PathBuf};

#[derive(Clone)]
pub struct Paths {
    pub data_dir: PathBuf,
}

impl Paths {
    pub fn new(data_dir: PathBuf) -> Self {
        let p = Self { data_dir };
        p.ensure();
        p
    }
    pub fn config(&self) -> PathBuf {
        self.data_dir.join("config.json")
    }
    pub fn listings(&self) -> PathBuf {
        self.data_dir.join("listings.json")
    }
    pub fn queue(&self) -> PathBuf {
        self.data_dir.join("queue.json")
    }
    pub fn db(&self) -> PathBuf {
        self.data_dir.join("wb.db")
    }
    pub fn images(&self) -> PathBuf {
        self.data_dir.join("images")
    }
    pub fn ensure(&self) {
        let _ = fs::create_dir_all(&self.data_dir);
        let _ = fs::create_dir_all(self.images());
    }
}

/// Write to a temp file then rename (same-dir rename is atomic), so a crash
/// mid-write can't truncate the real file.
pub fn atomic_write(file: &Path, data: &[u8]) -> std::io::Result<()> {
    if let Some(dir) = file.parent() {
        fs::create_dir_all(dir)?;
    }
    let mut tmp = file.as_os_str().to_owned();
    tmp.push(format!(".tmp.{}", std::process::id()));
    let tmp = PathBuf::from(tmp);
    fs::write(&tmp, data)?;
    fs::rename(&tmp, file)?;
    Ok(())
}

/// Move a corrupt/unparseable file aside instead of silently overwriting it.
pub fn quarantine_corrupt(file: &Path) {
    if file.exists() {
        let ts = chrono::Utc::now().timestamp_millis();
        let mut bak = file.as_os_str().to_owned();
        bak.push(format!(".corrupt-{}", ts));
        let _ = fs::rename(file, PathBuf::from(bak));
    }
}
